cdap-proto/src/main/java/io/cdap/cdap/proto/TriggeringInfo.java

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.proto;
+
+import io.cdap.cdap.api.schedule.Trigger;
+import io.cdap.cdap.proto.id.ProgramRunId;
+import io.cdap.cdap.proto.id.ScheduleId;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+/**
+ * Class contains the triggering information for a run
+ */
+public abstract class TriggeringInfo implements Trigger {
+  private final Type type;
+  private final ScheduleId scheduleId;
+  /**
+   * Runtime Arguments propagated from upstream program or properties of time schedule
+   */
+  @Nullable
+  private final Map<String, String> runtimeArguments;
+
+  protected TriggeringInfo(Type type, ScheduleId scheduleId, @Nullable Map<String, String> runtimeArguments) {
+    this.type = type;
+    this.scheduleId = scheduleId;
+    this.runtimeArguments = Collections.unmodifiableMap(runtimeArguments);
+  }
+
+  @Override
+  public Type getType() {
+    return type;
+  }
+
+  public ScheduleId getScheduleId() {
+    return scheduleId;
+  }
+
+  @Nullable
+  public Map<String, String> getRuntimeArguments() {
+    return runtimeArguments;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof TriggeringInfo)) {
+      return false;
+    }
+    TriggeringInfo that = (TriggeringInfo) o;
+    return Objects.equals(getType(), that.getType())
+      && Objects.equals(getScheduleId(), that.getScheduleId())
+      && Objects.equals(getRuntimeArguments(), that.getRuntimeArguments());
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(getType(), getScheduleId(), getRuntimeArguments());
+  }
+
+  /**
+   * Represents Triggering info of a time trigger
+   */
+  public static class TimeTriggeringInfo extends TriggeringInfo {
```

Comment:
can you explain why we need to create all of these new classes? Aren't there already Trigger objects? What is missing from them?